### PR TITLE
DAOS-9818 control: Component interoperability checks

### PR DESCRIPTION
**Pull request by mjmac**
Add a framework for performing interoperability checks
between components, and start using it to enforce
interoperability rules.

Current rules:
  * 2.0.0 and above are not backward-compatible with < 2.0.0
  * Components must be patch-compatible (e.g. 2.0.1 <-> 2.0.2)

Signed-off-by: Michael MacDonald <mjmac.macdonald@intel.com>

### CODE REVIEW

File: src/control/build/interop.go
Comment (knard-intel) [inline src/control/build/interop.go:73]: I am probably missing something with golang, but it seems to be prettier like this ?
```suggestion
	return fmt.Sprintf("%s:%s", vc.Component, vc.Version.String())
```

File: src/control/server/faults.go
Comment (knard-intel) [inline src/control/server/faults.go:142]: I am probably missing something with golang, but it seems to be prettier like this ?
```suggestion
		fmt.Sprintf("components %s and %s are not compatible", self.String(), other.String()),
```

File: src/control/build/interop.go
Comment (knard-intel) [inline src/control/build/interop.go:45]: Do not know if needed, but the component `"undefined"` should maybe added to have all the components from grpc_authorization.go ?

File: src/control/server/interceptors.go
Comment (mjmac) [reply src/control/server/interceptors.go:138]: Agreed. Unfortunately we don't have a good way to inject logging into the gRPC interceptor functions.

Comment (knard-intel) [reply src/control/server/interceptors.go:138]: OK

File: src/control/build/interop.go
Comment (mjmac) [reply src/control/build/interop.go:73]: Can you explain what you mean by "prettier"? Do you mean the code or the output? Currently, the stringer results in output like this: `incompatible components: server:1.0.0, server:1.1.0`

**Comment by mjmac** [comment]
FYI, I'm planning to re-run this with the `Features: control` pragma before landing, but I'll hold off on that until I get some more reviews.

File: src/control/build/interop.go
Comment (mjmac) [reply src/control/build/interop.go:73]: Not black magic exactly, just a format rule: https://pkg.go.dev/fmt#Stringer

File: src/control/build/interop.go
Comment (mjmac) [reply src/control/build/interop.go:45]: I don't think so... In the event that an undefined component makes it through it will correctly trigger an error in `NewVersionedComponent()`.

Comment (knard-intel) [reply src/control/build/interop.go:45]: OK

**Review by knard-intel — APPROVED** [review]
Fine for me

File: src/control/lib/control/request.go
Comment (mjmac) [reply src/control/lib/control/request.go:99]: I considered that. It would be clearer, but on the other hand it bloats the messages even more. I think I'd like to go forward with this approach unless you see a serious downside to it. We can always add version later, or find some other scheme to handle versioning.

One benefit of using the system field this way is that it guarantees that newer components can't talk to older servers without versioning support, because the system name check will fail (e.g. newer agent/dmg will send `system_foo-2.2.0` and that won't match `system_foo` on the server, so the request will be rejected).

File: src/control/build/interop.go
Comment (mjmac) [reply src/control/build/interop.go:152]: Yeah, I was trying to leave things somewhat flexible to handle non-standard requirements, but maybe it would be best to keep these hard-coded rules as simple as possible.

Another thought that occurred to me as I was re-reviewing this the other day was that I could remove these hard-coded rules altogether and just add them as the last set of default flexible rules.

Comment (kjacque) [reply src/control/build/interop.go:152]: If we need to accommodate multiple major versions at some point, the checks will probably get more complex than this anyway.

I think it would make sense to include them in flexible rules.

File: src/control/lib/control/request.go
Comment (mjmac) [reply src/control/lib/control/request.go:99]: > this would have to be backported to 2.0 in order to support 2.0-2.2 compatibility, right?

Nope. We don't guarantee that 2.2 components should be able to talk to 2.0 servers.

Comment (mjmac) [reply src/control/lib/control/request.go:99]: > But I think either way, the request needs a getVersion() method to make the world simpler for the callers using it.

Yes, this is a good idea. I'll incorporate that in the next push.

**Review by tanabarr — APPROVED** [review]

File: src/control/build/interop.go
Comment (kjacque) [inline src/control/build/interop.go:77]: Unclear to me what purpose these two members serve. Is the idea to have different rules for agent to server, server to agent, etc?

Comment (kjacque) [inline src/control/build/interop.go:77]: Do these two members still serve a purpose?

**Review by kjacque — APPROVED** [review]

**Review by kjacque — APPROVED** [review]

File: src/control/build/interop.go
Comment (mjmac) [reply src/control/build/interop.go:77]: Yes, exactly. At the moment we're not using them because the interop rules are very simple and only version-based. In a previous iteration I had more complex rules for checking server<->server and agent<->server compatibility, but those aren't necessary right now.

**Pull request merged**
